File: UnitTesting/Old/SpecFlow-Examples-master/BowlingKata/BowlingKata-VB-MsTest/Bowling/Game.vb
Namespace Bowling
    Public Class Game
        Private rolls As Integer() = New Integer(21) {}
        Private currentRoll As Integer

        Public Sub Roll(ByVal pins As Integer)
            currentRoll = currentRoll + 1
            rolls(currentRoll) = pins
        End Sub

        Public ReadOnly Property Score() As Integer
            Get
                Dim theScore As Integer = 0
                Dim frameIndex As Integer = 1
                For frame As Integer = 1 To 10
                    If isStrike(frameIndex) Then
                        theScore += 10 + strikeBonus(frameIndex)
                        frameIndex += 1
                    ElseIf isSpare(frameIndex) Then
                        theScore += 10 + spareBonus(frameIndex)
                        frameIndex += 2
                    Else
                        theScore += sumOfBallsInFrame(frameIndex)
                        frameIndex += 2
                    End If
                Next
                Return theScore
            End Get
        End Property

        Private Function isStrike(ByVal frameIndex As Integer) As Boolean
            Return rolls(frameIndex) = 10
        End Function

        Private Function sumOfBallsInFrame(ByVal frameIndex As Integer) As Integer
            Return rolls(frameIndex) + rolls(frameIndex + 1)
        End Function

        Private Function spareBonus(ByVal frameIndex As Integer) As Integer
            Return rolls(frameIndex + 2)
        End Function

        Private Function strikeBonus(ByVal frameIndex As Integer) As Integer
            Return rolls(frameIndex + 1) + rolls(frameIndex + 2)
        End Function

        Private Function isSpare(ByVal frameIndex As Integer) As Boolean
            Return rolls(frameIndex) + rolls(frameIndex + 1) = 10
        End Function

    End Class
End Namespace

File: UnitTesting/Old/SpecFlow.Bowling/Game.cs
namespace Bowling.Specflow
{
    public class Game
    {
        private readonly int[] _rolls = new int[21];
        private int _currentRoll;

        public void Roll(int pins)
        {
            _rolls[_currentRoll++] = pins;
        }

        public int Score
        {
            get
            {
                int score = 0;
                int frameIndex = 0;
                for (int frame = 0; frame < 10; frame++)
                {
                    if (isStrike(frameIndex))
                    {
                        score += 10 + strikeBonus(frameIndex);
                        frameIndex++;
                    }
                    else if (isSpare(frameIndex))
                    {
                        score += 10 + spareBonus(frameIndex);
                        frameIndex += 2;
                    }
                    else
                    {
                        score += sumOfBallsInFrame(frameIndex);
                        frameIndex += 2;
                    }
                }
                return score;
            }
        }

        private bool isStrike(int frameIndex)
        {
            return _rolls[frameIndex] == 10;
        }

        private int sumOfBallsInFrame(int frameIndex)
        {
            return _rolls[frameIndex] + _rolls[frameIndex + 1];
        }

        private int spareBonus(int frameIndex)
        {
            return _rolls[frameIndex + 2];
        }

        private int strikeBonus(int frameIndex)
        {
            return _rolls[frameIndex + 1] + _rolls[frameIndex + 2];
        }

        private bool isSpare(int frameIndex)
        {
            return _rolls[frameIndex] + _rolls[frameIndex + 1] == 10;
        }
    }
}
